### PR TITLE
Orchestrator examples

### LINKED ISSUE #67: Add example: Integration with workflow engines
Integration with workflow engines, structuring into a benchmarking task. (E.g. flyte, prefect, dagster, Beam, ...).

### PR DESCRIPTION
**Pull request by maxmynter**
Add Prefect example

To test it, you have to do the following:

- Install Prefect: `pip install -U prefect`
- Start prefect Server: `prefect server start`
- Expose the benchmark flows by executing the `runner.py` in the `examples/prefect` directory: `python runner.py`
- Then you can trigger an execution of the exposed flow by: `prefect deployment run 'train-and-benchmark/benchmark-runner'`

The default local UI of Prefect is `localhost:4200`.

Closes #67

### CODE REVIEW

File: examples/prefect/model_data_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import pickle
```
Comment (nicholasjng) [inline examples/prefect/model_data_benchmark.py:1]: I would say merge the files into one - the tags provide acceptable semantic grouping already.

File: examples/prefect/runner.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    def write(
+        self,
+        record: types.BenchmarkRecord,
+        benchmark_filter: str | None = None,
+        include_context: tuple[str, ...] = (),
+        exclude_empty: bool = True,
+    ) -> None:
+        old_stdout = sys.stdout
+        sys.stdout = StringIO()
```
Comment (nicholasjng) [inline examples/prefect/runner.py:26]: oje, you should never monkey-patch stdlib functions in code.

The way to go here is to attach a console handler to the obtained logger - see the documentation of the logging module on handlers: https://docs.python.org/3/library/logging.html#handler-objects

File: examples/prefect/training.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+def make_train_test_split(
+    X: np.ndarray, y: np.ndarray, random_state: int, test_size: float = 0.2
+) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    X_train, X_test, y_train, y_test = train_test_split(
+        X, y, test_size=test_size, random_state=random_state
+    )
+    return X_train, y_train, X_test, y_test
+
+
+@flow
```
Comment (nicholasjng) [inline examples/prefect/training.py:29]: Do we ever need to prepare regression data outside of the flow afterwards? If not, we could think about inlining this into `prepare_regressor_and_test_data` below.

Is there any noticeable advantage of having it on the dashboard etc. in your opinion?

File: examples/prefect/training.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    X_train, y_train, X_test, y_test = prepare_regression_data(random_state=random_state)
+    model = train_linear_regression(X=X_train, y=y_train)
+    return model, X_test, y_test
+
+
+@task
+def train_linear_regression(X: np.ndarray, y: np.ndarray) -> base.BaseEstimator:
+    model = LinearRegression()
+    model.fit(X, y)
+    return model
```
Comment (nicholasjng) [inline examples/prefect/training.py:51]: Another candidate for inlining. (Also please define all tasks after the flows, seems to be a convention for Python orchestrators in general.)

File: examples/prefect/model_data_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    return sys.getsizeof(model_bytes)
+
+
+@nnbench.benchmark(name="Inference time (s)", tags=("model-meta",))
+def inference_time(model: base.BaseEstimator, X: np.ndarray, n_iter: int = 100) -> float:
+    start = time.perf_counter()
+    for i in range(n_iter):
+        _ = model.predict(X)
+    end = time.perf_counter()
+    return (end - start) / n_iter
```
Comment (nicholasjng) [inline examples/prefect/model_data_benchmark.py:23]: We time each benchmark anyway, so we technically have this in the bag. Any idea how could we give the option to use the execution time as the target metric instead?

(Otherwise, it's fine - though the `time_ns` metric then becomes superfluous.)

File: examples/prefect/model_data_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import numpy as np
+from sklearn import base
+
+import nnbench
+
+
+@nnbench.benchmark(name="Model size (bytes)", tags=("model-meta",))
+def modelsize(model: base.BaseEstimator) -> int:
+    model_bytes = pickle.dumps(model)
+    return sys.getsizeof(model_bytes)
```
Comment (nicholasjng) [inline examples/prefect/model_data_benchmark.py:14]: Is there any reason to not just call `sys.getsizeof()` on the model directly?

File: examples/prefect/benchmark.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import numpy as np
+from prefect import task
+from sklearn import base, metrics
+
+import nnbench
+
+
+@nnbench.benchmark
```
Comment (nicholasjng) [inline examples/prefect/benchmark.py:8]: I would leave this file out until we can think of a way to facilitate the nnbench.Benchmark<->prefect.Task correspondence.

File: examples/prefect/model_data_benchmark.py (hunk 6 above)
Comment (maxmynter) [reply examples/prefect/model_data_benchmark.py:14]: Calling on the object directly does only give the memory footprint of the object itself, not of any other values it references. See [here](https://stackoverflow.com/questions/19516403/why-pickle-dumpobj-has-different-size-with-sys-getsizeofobj-how-to-save-var). Thats why I went with serializing first.

File: examples/prefect/model_data_benchmark.py (hunk 5 above)
Comment (maxmynter) [reply examples/prefect/model_data_benchmark.py:23]: Good point. I'll think of another metadata metric we can use.

File: examples/prefect/benchmark.py (hunk 7 above)
Comment (maxmynter) [reply examples/prefect/benchmark.py:8]: Damn, yeah. I deleted it but it must have somehow come back  while squashing when i was cleaning up the git history.

File: examples/prefect/training.py (hunk 3 above)
Comment (maxmynter) [reply examples/prefect/training.py:29]: I was kinda on the fence about this one and had it inline at first. 
It makes a nice dependency graph in the Prefect UI but we are also not demo'ing prefect and the difference is not that big. Factoring it back

File: examples/prefect/training.py (hunk 4 above)
Comment (maxmynter) [reply examples/prefect/training.py:51]: The thing is if we inline everything the example may become too small to show the integration into Prefect.

And you mean flows after tasks, right? At least thats what the Prefect docs do.

File: examples/prefect/runner.py (hunk 2 above)
Comment (maxmynter) [reply examples/prefect/runner.py:26]: Mhhh that does not seem to work. Prefect supplies a `LoggerAdapter` not a `Logger`. To the former one cannot attach streams.

Comment (maxmynter) [reply examples/prefect/runner.py:26]: Decided on leaving out formatted reporting. In a real life scenario reports would go into a database and would then be read from a dedicated UI, not the orchestrator. 

I will implement a barebones reporter that simply logs the `dict` returned by the benchmarks to the console though.

File: examples/prefect/model_data_benchmark.py (hunk 5 above)
Comment (maxmynter) [reply examples/prefect/model_data_benchmark.py:23]: Though the `time_ns ` times the whole benchmark. Here we know the time it takes for the plain inference. Both are different (esp. with more complex benchmarks) but of interest, no?

Comment (nicholasjng) [reply examples/prefect/model_data_benchmark.py]: Yeah, what I meant is that both should be the same if the benchmark consists only of "_ = model.predict(X)".

But that's statistically unstable, so please leave it as is.

File: examples/prefect/training.py (hunk 4 above)
Comment (nicholasjng) [reply examples/prefect/training.py:51]: Yes, flows after tasks, sorry.

File: docs/tutorials/prefect.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
+# Integrating nnbench with Prefect
+
+If you have more complex workflows it is sensible to use a workflow orchestration tools. 
+Benchmarking with nnbench can be integrated with orchestrators. We will present an example integration with Prefect.
+We will explain the orchestration concepts in a high level and link to the corresponding parts of the  Prefect [docs](https://docs.prefect.io/).
+The full example code is in the nnbench [repository](https://github.com/aai-institute/nnbench/tree/main/examples/prefec).
```
Comment (nicholasjng) [inline docs/tutorials/prefect.md:6]:
```suggestion
The full example code can be found in the nnbench [repository](https://github.com/aai-institute/nnbench/tree/main/examples/prefect).
```

File: docs/tutorials/prefect.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
+# Integrating nnbench with Prefect
+
+If you have more complex workflows it is sensible to use a workflow orchestration tools.
```
Comment (nicholasjng) [inline docs/tutorials/prefect.md:3]:
```suggestion
If you have more complex workflows, it is sensible to use a workflow orchestration tool to manage them. 
```

File: docs/tutorials/prefect.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+## Project Structure
+### Defining the training tasks and workflows
+We recommend to separate the training and benchmarking logic. 
+
+```python
+--8<-- "examples/prefect/training.py"
+```
+
+The `training.py` file contains functions to generate synthetic data for our regression model, facilitate a train-test-split, and finally train the regression model.
+We have applied Prefects [`@task` decorator.](https://docs.prefect.io/latest/concepts/tasks/)
```
Comment (nicholasjng) [inline docs/tutorials/prefect.md:19]:
```suggestion
We have applied Prefect's [`@task` decorator.](https://docs.prefect.io/latest/concepts/tasks/)  
```

I would also merge the next sentence into it, e.g. like "which marks the function logic ..."

File: docs/tutorials/prefect.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+```python
+--8<-- "examples/prefect/training.py"
+```
+
+The `training.py` file contains functions to generate synthetic data for our regression model, facilitate a train-test-split, and finally train the regression model.
+We have applied Prefects [`@task` decorator.](https://docs.prefect.io/latest/concepts/tasks/)  
+It marks the contained logic as a discrete unit of work for Prefect. 
+Two other functions prepare the regression data and train the estimator. 
+They are labeled with the [`@flow` decorator.](https://docs.prefect.io/latest/concepts/flows) which labels the function as a workflow logic container that can depend on other flows or tasks.
```
Comment (nicholasjng) [inline docs/tutorials/prefect.md:22]:
```suggestion
They are labeled with the [`@flow` decorator.](https://docs.prefect.io/latest/concepts/flows) which labels the function as a workflow that can depend on other flows or tasks.
```

File: docs/tutorials/prefect.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+The benchmarks are in the `benchmark.py` file.
+We have two functions to calculate the mean absolute error and the mean squared error.
+These benchmarks are tagged to indicate they are metrics.
+The other two benchmarks calculate model metadata, namely the size and inference time of the model.
+
+```python
+--8<-- "examples/prefect/benchmark.py"
+```
+
+We did not apply any Prefect decorators here as they are incompatible with nnbench's decorators.
```
Comment (nicholasjng) [inline docs/tutorials/prefect.md:35]: True, but they are also a finer-grained unit of work, as we currently operate under the "one benchmark family per task" paradigm. I would mention that here instead.

**Review by nicholasjng — COMMENTED** [review]
Sweet!
